### PR TITLE
Filter-by-key code timing fixed, openMP flag removed.

### PR DESCRIPTION
**Pull request by fzx333**
Hi @deyuan @fasiddique, I have completely updated the filter-by-key code. Now both the PIM code and the CPU don't use openMP anymore, I've also updated the makefile. I have also tested the code, and the code works fine. Can you also check the make file and see if everything looks good to you and merge the changes?

### CODE REVIEW

**Comment by fasiddique** [comment]
@fzx333 there's some conflicts will you please resolve it first

**Comment by fzx333** [comment]
@fasiddique, I saw that. Can I delete this push and this PR? I will create a new push.

**Comment by fasiddique** [comment]
> @fasiddique, I saw that. Can I delete this push and this PR? I will create a new push.

Sure!

**Comment by fzx333** [comment]
Hi @fasiddique, I just forced another push; now this PR has no conflict! You can review it now. Thanks.

File: PIMbench/filter-by-key/PIM/filter.cpp [hunk 1]
```diff
@@ -9,26 +9,27 @@
...
 
-#define MY_RANGE 100
+#include <immintrin.h>
+
+#define MY_RANGE 1000
 
 using namespace std;
 
-std::chrono::duration<double, std::milli> hostElapsedTime = std::chrono::duration<double, std::milli>::zero();
+// std::chrono::duration<double, std::milli> hostElapsedTime = std::chrono::duration<double, std::milli>::zero();
```
Comment (fasiddique) [inline PIMbench/filter-by-key/PIM/filter.cpp:24]:
```suggestion
```
cleaning up

File: PIMbench/filter-by-key/PIM/filter.cpp [hunk 2]
```diff
@@ -9,26 +9,27 @@
 #include <vector>
 #include <getopt.h>
 #include <cstdlib>
-#include <time.h>
+#include <time.h> 
 #include <algorithm>
 #include <chrono>
-#if defined(_OPENMP)
 #include <omp.h>
-#endif
```
Comment (fasiddique) [inline PIMbench/filter-by-key/PIM/filter.cpp:17]: I think this is necessary to build successfully without openmp

File: PIMbench/filter-by-key/baselines/CPU/filter.cpp [hunk 3]
```diff
@@ -13,39 +7,44 @@
...
 
-#define MY_RANGE 100
+#define MY_RANGE 1000
 
 using namespace std;
 
+// struct Record {
+//   int32_t value1;
+//   int32_t value2;
+// };
```
Comment (fasiddique) [inline PIMbench/filter-by-key/baselines/CPU/filter.cpp:21]:
```suggestion
```

File: PIMbench/filter-by-key/baselines/CPU/filter.cpp [hunk 4]
```diff
@@ -13,39 +7,44 @@
...
-}
+// void filterByKey(std::vector<Record> &Vector, uint64_t vector_size, int key, std::vector<bool> & bitMap)
+// {
+// #pragma omp parallel for
+//   for (uint64_t i = 0; i < vector_size; ++i)
+//   {
+//     if (key > Vector[i].value1 && key > Vector[i].value2)
+//       bitMap[i] = true;
+//   }
+// }
```
Comment (fasiddique) [inline PIMbench/filter-by-key/baselines/CPU/filter.cpp:35]:
```suggestion
```

File: PIMbench/filter-by-key/baselines/CPU/filter.cpp [hunk 5]
```diff
@@ -13,39 +7,44 @@
...
-    int key;
+    uint64_t key;
     bool shouldVerify;
 } Params;
 
 void usage()
 {
   fprintf(stderr,
-          "\nUsage:  ./filter.out [options]"
+          "\nUsage:  ./filter [options]"
```
Comment (fasiddique) [inline PIMbench/filter-by-key/baselines/CPU/filter.cpp:47]: shouldn't it have .out?

File: PIMbench/filter-by-key/PIM/filter.cpp (hunk 2 above)
Comment (fzx333) [reply PIMbench/filter-by-key/PIM/filter.cpp:17]: Sorry, this "#include <omp.h>" is actually no long needed, can you just help me remove that, I didn't notice that.

File: PIMbench/filter-by-key/baselines/CPU/filter.cpp (hunk 5 above)
Comment (fzx333) [reply PIMbench/filter-by-key/baselines/CPU/filter.cpp:34]: Yes, it should have .out!

File: PIMbench/filter-by-key/PIM/filter.cpp [hunk 6]
```diff
@@ -86,160 +87,152 @@ struct Params getInputParams(int argc, char **argv)
...
     }
-  }
-  pimShowStats();
+    uint64_t outSize = selectedNum1 + selectedNum2 + selectedNum3 + selectedNum4 + selectedNum5 + selectedNum6 + selectedNum7 + selectedNum8;
 
-  cout << "Host elapsed time: " << std::fixed << std::setprecision(3) << hostElapsedTime.count() << " ms." << endl;
-  return 0;
-}
+    auto stop_cpu = std::chrono::high_resolution_clock::now();
+    std::chrono::duration<double, std::milli> hostElapsedTime = (stop_cpu - start_cpu);
```
Comment (fasiddique) [inline PIMbench/filter-by-key/PIM/filter.cpp:231]: Is it possible to move this code block to a function call?

Comment (fzx333) [reply PIMbench/filter-by-key/PIM/filter.cpp:231]: It is a performance-specific optimization, I think it is safer to just keep it locally in the main function, and I think the code also runs faster in this way.

**Comment by fzx333** [comment]
Hi @fasiddique, now I completely removed the "#include <omp.h>" in the file, and it gets compiled fine. I have also changed the .out in the prompt!

File: PIMbench/filter-by-key/PIM/filter.cpp [hunk 7]
```diff
@@ -86,160 +85,152 @@ struct Params getInputParams(int argc, char **argv)
...
+    if (p.inputFile == nullptr){
+        inVectorSize = p.inVectorSize;
+        key = p.key;
+    }
+    else{
+        std::cout << "Reading from input file is not implemented yet." << std::endl;
+        return 1;
+    }
+    bool* bitMapTemp = new bool[inVectorSize];
+    // vector<bool> bitMapTemp(inVectorSize);
```
Comment (fasiddique) [inline PIMbench/filter-by-key/PIM/filter.cpp:105]:
```suggestion
```

File: PIMbench/filter-by-key/PIM/filter.cpp [hunk 8]
```diff
@@ -86,160 +85,152 @@ struct Params getInputParams(int argc, char **argv)
...
-    bool ok = true;
-    if (outVector != outVectorHost)
-    {
-      std::cout << "Wrong answer!" << std::endl;
-      ok = false;
-    }
-    if (ok)
-    {
-      std::cout << "All correct!" << std::endl;
+    // // uint64_t outSize = selectedNum;
```
Comment (fasiddique) [inline PIMbench/filter-by-key/PIM/filter.cpp:172]:
```suggestion
```

File: PIMbench/filter-by-key/baselines/CPU/filter.cpp [hunk 9]
```diff
@@ -90,78 +76,53 @@ struct Params getInputParams(int argc, char **argv)
...
+    uint64_t dummyVectorSize = 1073741824;
+    vector<int> dummyVector1(dummyVectorSize, 0);
+    uint64_t buffer_in_CPU = 0;
+    uint64_t selectedNum = 0;
+
+    // Flushing the cache
+    for (uint64_t j = 0; j < dummyVectorSize; j++){
+      dummyVector1[j] += rand() % MY_RANGE;
     }
+    // cout << "Cache flushed!" << endl;
```
Comment (fasiddique) [inline PIMbench/filter-by-key/baselines/CPU/filter.cpp:106]:
```suggestion
```

File: PIMbench/filter-by-key/baselines/CPU/filter.cpp [hunk 10]
```diff
@@ -90,78 +76,53 @@ struct Params getInputParams(int argc, char **argv)
...
+      if(p.key > inVector[i]){
+        buffer_in_CPU += inVector[i];
+        selectedNum++;
+      }
+    }
+    uint64_t outSize = selectedNum;
     auto end = std::chrono::high_resolution_clock::now();
-    std::chrono::duration<double, std::milli> elapsedTime = (end - start);
+    elapsedTime = (end - start);
+    // cout << "buffer_in_CPU = " << buffer_in_CPU << endl;
```
Comment (fasiddique) [inline PIMbench/filter-by-key/baselines/CPU/filter.cpp:119]:
```suggestion
```

**Review by fasiddique — APPROVED** [review]
thanks! LGTM

**Pull request merged**